provider_client.go

```diff
@@ -126,13 +142,53 @@ func (client *ProviderClient) Token() string {
 }
 
 // SetToken safely sets the value of the auth token in the ProviderClient. Applications may
-// use this method in a custom ReauthFunc
+// use this method in a custom ReauthFunc.
+//
+// WARNING: This function is deprecated. Use SetTokenFromAuthResult() instead.
 func (client *ProviderClient) SetToken(t string) {
 	if client.mut != nil {
 		client.mut.Lock()
 		defer client.mut.Unlock()
 	}
 	client.TokenID = t
+	client.authResult = nil
+}
+
+//SetTokenFromAuthResult safely sets the value of the auth token in the ProviderClient
+//and also records the AuthResult that was returned from the token creation
+//request. Applications may use this method in a custom ReauthFunc.
+func (client *ProviderClient) SetTokenFromAuthResult(r AuthResult) error {
```

Comment:
`SetTokenFromAuthResult` is also setting `client.authResult`. The longest part of me re-reviewing this was figuring out how `authResult` was being set. It would have been more immediately clear to me, and so most likely others, if the name was `SetTokenAndAuthResult`. Or have two separate methods: one to set the token and one to set the auth result.

Reply:
I wouldn't want two separate methods, so I renamed to `SetTokenAndAuthResult`.